rflx/pyrflx/typevalue.py

```diff
@@ -606,22 +612,49 @@ def __eq__(self, other: object) -> bool:
     def equal_type(self, other: Type) -> bool:
         return self.identifier == other.identifier
 
+    @property
+    def structure(self) -> Sequence[Link]:
+        return self._type.structure
+
+    @property
+    def path(self) -> Sequence[Link]:
+        return self._path
+
+    def inner_messages(self) -> List["MessageValue"]:
+        sdu_messages: List["MessageValue"] = []
+        for field in self._fields.values():
+            typeval = field.typeval
+            if isinstance(typeval, ArrayValue) and isinstance(typeval.element_type, Message):
+                assert isinstance(typeval.value, list)
+                sdu_messages.extend(typeval.value)
+            if isinstance(typeval, OpaqueValue) and typeval.nested_message is not None:
+                sdu_messages.append(typeval.nested_message)
+        return sdu_messages
+
     def _valid_refinement_condition(self, refinement: "RefinementValue") -> bool:
         return self.__simplified(refinement.condition) == TRUE
 
-    def _next_field(self, fld: str) -> str:
-        if fld == FINAL.name:
-            return ""
-        if self._skip_verification and self._fields[fld].next:
-            return self._fields[fld].next
-        if fld == INITIAL.name:
-            links = self._type.outgoing(INITIAL)
-            return links[0].target.name if links else FINAL.name
+    def _next_link(self, source_field_name: str) -> Optional[Link]:
+        field = Field(source_field_name)
+        # structure is empty for NULL messages
+        if field == FINAL or not self._type.structure:
+            return None
+        for link in self._type.outgoing(field):
+            if self.__simplified(link.condition) == TRUE:
+                return link
+        return None
 
-        for l in self._type.outgoing(Field(fld)):
-            if self.__simplified(l.condition) == TRUE:
-                return l.target.name
-        return ""
+    def _next_field(self, field_name: str, coverage: bool = False) -> str:
+        if self._skip_verification and field_name != FINAL.name and self._fields[field_name].next:
+            return self._fields[field_name].next
+
+        next_link = self._next_link(field_name)
+        if coverage and next_link is not None:
+            self._path.append(next_link)
```

Comment:
Setting field in non-sequential order using the `set()` method would currently result in an empty `path`. It would also probably not work correctly if the `skip_verification` flag is set.
The parser sets all fields in sequential order and just once, so the current implementation would be sufficient for the use-case we had in mind in the validation tool. 

Another idea that I had for storing the path/coverage in the first place is to not use the boolean flag as discussed and also not "save" the used link every time a field is set.  Instead, the `path` method would iterate through all fields that are `set` and then reconstruct the used links from the set fields. That should work since there are no parallel links between fields and all `set` fields are connected. Using this approach, it would not matter if the `set` or `parse` method is used or the `skip_verification` flag is set.

Reply:
> Setting field in non-sequential order using the `set()` method would currently result in an empty `path`.

Can we ensure that `path` is empty when `set()` is used manually? What happens with `path` when `parse` is used more than once?

> It would also probably not work correctly if the `skip_verification` flag is set.

That shouldn't be an issue, but should be documented.

> the `path` method would iterate through all fields that are `set` and then reconstruct the used links from the set fields. That should work since there are no parallel links between fields and all `set` fields are connected. Using this approach, it would not matter if the `set` or `parse` method is used or the `skip_verification` flag is set.

How expensive would that recalculation be? Although I would prefer a `path` function which works in all cases, a unnecessary slow validation tool would be worse. If the recalculation is expensive, we should rather document the limitations of the `path` function and raise an exception in `path` for non-supported use cases.

Reply:
I profiled the `test_validation_negative_full_output` unit test for the validation tool. There are in total 15 calls to `path` during execution of the test.
Here is the result:

alternative 1 := recalculation of the path
alternative 2 := storing the used links during parsing

execution time of `path` with alternative 1: 771 ms (5.2 % of total execution time)
execution time of `path` with alternative 2: 0 ms

total execution time with alternative 1: 2484 ms
total execution time with alternative 2: 1696 ms

Reply:
>Can we ensure that path is empty when set() is used manually? What happens with path when parse is used more than once?

The flag to store the used link is only passed in calls to `next_field` from inside `parse`. So if only `set()` is used to manually set fields, `path` would be guaranteed to be empty. The problem is when first `parse` is used and the message is manipulated with `set()` afterwards. In that case `path` would be incorrect. It is also not possible to simply clear `path` when `set()` is called, as the `parse` method calls `set()`.

>What happens with path when parse is used more than once?

Currently, the new path would be appended to the old one, but that is easy to fix. When calling `parse`, `path` can be cleared first.